### DISCUSSION TITLE
Error when running rules on custom components

### DISCUSSION THREAD

**Discussion by Brannos1970** [MarimerLLC/csla#1780] [Questions]
I am getting the following error when updating a field in my Blazor application:

> crit: Microsoft.AspNetCore.Components.WebAssembly.Rendering.WebAssemblyRenderer[100]
>       Unhandled exception rendering component: Model does not implement required interface ICheckRules
> System.ArgumentException: Model does not implement required interface ICheckRules
>   at Csla.Blazor.EditContextCslaExtensions.ValidateField (Microsoft.AspNetCore.Components.Forms.EditContext editContext, Microsoft.AspNetCore.Components.Forms.ValidationMessageStore messages, Microsoft.AspNetCore.Components.Forms.FieldIdentifier& fieldIdentifier) <0x4a91e48 + 0x00030> in <filename unknown>:0 
>   at Csla.Blazor.EditContextCslaExtensions+<>c__DisplayClass0_0.<AddCslaValidation>b__1 (System.Object sender, Microsoft.AspNetCore.Components.Forms.FieldChangedEventArgs eventArgs) <0x4a91b80 + 0x0002a> in <filename unknown>:0 
>   at (wrapper delegate-invoke) System.EventHandler`1[Microsoft.AspNetCore.Components.Forms.FieldChangedEventArgs].invoke_void_object_TEventArgs(object,Microsoft.AspNetCore.Components.Forms.FieldChangedEventArgs)
>   at Microsoft.AspNetCore.Components.Forms.EditContext.NotifyFieldChanged (Microsoft.AspNetCore.Components.Forms.FieldIdentifier& fieldIdentifier) <0x4a90cb0 + 0x00046> in <filename unknown>:0 
>   at Radzen.Blazor.RadzenTextBox.OnChange (Microsoft.AspNetCore.Components.ChangeEventArgs args) <0x4a60eb0 + 0x00170> in <filename unknown>:0 
>   at Microsoft.AspNetCore.Components.ComponentBase.CallStateHasChangedOnAsyncCompletion (System.Threading.Tasks.Task task) <0x347e930 + 0x000ee> in <filename unknown>:0 
>   at Microsoft.AspNetCore.Components.RenderTree.Renderer.GetErrorHandledTask (System.Threading.Tasks.Task taskToHandle) <0x32809f0 + 0x000ca> in <filename unknown>:0 

I thought this error was in the ViewModel but I realized it was coming from the BusinessBase. I checked the BB and the implementation is there. Is this an issue in CSLA I am not tracking down or did I miss something?

**Comment by rockfordlhotka** [comment]
I'd have to see your code (at least some of it) to understand.

* The viewmodel and/or the `@inject` of the viewmodel
* The UI binding code

**Reply by Brannos1970** [reply]
I have been tracing this down and found it was in my custom components. I changed the title to reflect this. Similar to the examples @rockfordlhotka provided for Blazor apps, I created custom components to simplify things and to standardized display.

In an Upsert razor page, I can enter the following (not all the code is shown for brevity as noted by the triple periods) and I get no errors when the value changes and the rules are run:

```
@inject Csla.Blazor.ViewModel<AppointmentEdit> VM
...
 <RadzenTextBox style="width: 100%" @bind-Value="@vm.Model.Summary" 
                   Name="@nameof(vm.Model.Summary)" 
                   AutoComplete="false"
                   Disabled="@(!vm.GetPropertyInfo(nameof(vm.Model.Summary)).CanWrite)"
                   Visible="@vm.GetPropertyInfo(nameof(vm.Model.Summary)).CanRead" />
 <span class="text-danger">@vm.GetPropertyInfo(nameof(vm.Model.Summary)).ErrorText</span>
 <span class="text-warning">@vm.GetPropertyInfo(nameof(vm.Model.Summary)).WarningText</span>
 <span class="text-info">@vm.GetPropertyInfo(nameof(vm.Model.Summary)).InformationText</span>
...
```
The AppointmentEdit class:
```
[Serializable]
    public class AppointmentEdit : BusinessBase<AppointmentEdit>
    {
        #region Properties
        ...
        public static readonly PropertyInfo<string> SummaryProperty = RegisterProperty<string>(nameof(Summary),  
                   GlobalResource.Summary);
        [Required]
        [Display(Name = nameof(GlobalResource.Summary), ResourceType = typeof(GlobalResource))]
        public string Summary
        {
            get => GetProperty(SummaryProperty);
            set => SetProperty(SummaryProperty, value);
        }
        ...
        #endregion

        #region Rules
        protected override void AddBusinessRules()
        {
            base.AddBusinessRules();
           
            BusinessRules.AddRule(new Csla.Rules.CommonRules.Required(SummaryProperty));
        }
        ...
        #endregion
        ...
    }
```
The above code works with no problems, I want to avoid entering code similar to this over and over so I created a custom component:
```
<div>
        <RadzenTextBox ReadOnly="@ReadOnly"
                       AutoComplete="false" 
                       Attributes="@AdditionalAttributes"
                       @bind-Value="@TextValue"
                       Name="@nameof(PropertyValue)" 
                       Disabled="@(!PropertyValue.CanWrite)"
                       Visible="@(!PropertyValue.CanRead)"/>
        <span class="text-danger">@PropertyValue.ErrorText</span>
        <span class="text-warning">@PropertyValue.WarningText</span>
        <span class="text-info">@PropertyValue.InformationText</span>
    </div>

@code {
    [Parameter(CaptureUnmatchedValues = true)]
    public Dictionary<string, object> AdditionalAttributes { get; set; }
    [Parameter]
    public IPropertyInfo PropertyValue { get; set; }
    [Parameter]
    public bool ReadOnly { get; set; }

    private string TextValue
    {
        get => (string)PropertyValue.Value;
        set => PropertyValue.Value = value;
    }
}
```
I then used this to replace the original code as:

`<TextBoxInput PropertyValue="@vm.GetPropertyInfo(nameof(vm.Model.Summary))" ReadOnly="false"  />`

When the value is changed I get the error. This is happening with all my custom controls that have properties with rules.

**Reply by rockfordlhotka** [reply]
That seems like it should work - looks quite similar to my components.

Can you provide the `ToString` of the exception? That'll provide more in-depth info about the exception stack.

**Comment by Brannos1970** [comment]
I changed the component to use <input> so the code looks like:

```
@if (visible)
{
    <div>
        <input readonly="@ReadOnly"
               autocomplete="off"
               class="textbox valid"
               style="width:100%"
               @attributes="@AdditionalAttributes"
               name="@nameof(PropertyValue)"
               @bind-value="TextValue"
               @bind-value:event="oninput"
               disabled="@(!PropertyValue.CanWrite)" />
        <br />
        <span class="text-danger">@PropertyValue.ErrorText</span>
        <span class="text-warning">@PropertyValue.WarningText</span>
        <span class="text-info">@PropertyValue.InformationText</span>
    </div>
}

@code {
    [Parameter(CaptureUnmatchedValues = true)]
    public Dictionary<string, object> AdditionalAttributes { get; set; }
    [Parameter]
    public Csla.Blazor.IPropertyInfo PropertyValue { get; set; }
    [Parameter]
    public bool ReadOnly { get; set; }

    private bool visible => PropertyValue.CanRead;

    private string TextValue
    {
        get => (string)PropertyValue.Value;
        set => PropertyValue.Value = value;
    }
}
```
The error went away. So I believe it is something with the Radzen component and how it renders.

I tried to get did switch back and try to get you more details but I was not able to capture more info. When I get a chance to further look at this I will and see if I can figure this out further and if other third-party custom components so the same thing.

**Comment by Brannos1970** [comment]
How do are we supposed to debug rules when we can't inject logging and the debugger won't hit breakpoints? I am finding several custom rules that are not working or throwing errors and I am wondering if that is part of my problem?

**Reply by rockfordlhotka** [reply]
Why won't the debugger hit breakpoints?

Also, have you considered creating unit tests for your rules? The best way to get rules working, in general, is to test them independently so you know they work, then attach them to properties as needed.

**Reply by Brannos1970** [reply]
I have not been able to figure out why the debugger is having issues hitting breakpoints.

I will get the tests set up and see if that helps me figure this out. Thanks @rockfordlhotka

**Reply by rockfordlhotka** [reply]
Sometimes I find that I have to do a solution rebuild (not build) to get everything in sync for the debugger.

**Reply by BlagoCuljak** [reply]
When debbuger is not hitting breakpoints, delete the hidden .vs folder in your solution.
